crates/sui-graphql-client/src/faucet.rs

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::time::Duration;
+
+use anyhow::bail;
+use reqwest::{StatusCode, Url};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use sui_types::types::{Address, ObjectId, TransactionDigest};
+
+use crate::{DEVNET_HOST, LOCAL_HOST};
+
+const FAUCET_DEVNET_HOST: &str = "https://faucet.devnet.sui.io/v1/gas";
+const FAUCET_TESTNET_HOST: &str = "https://faucet.testnet.sui.io/v1/gas";
+const FAUCET_LOCAL_HOST: &str = "http://localhost:9123/v1/gas";
+
+const FAUCET_REQUEST_TIMEOUT: Duration = Duration::from_secs(120);
+const FAUCET_POLL_INTERVAL: Duration = Duration::from_secs(2);
+
+pub struct FaucetClient {
+    faucet_url: Url,
+    inner: reqwest::Client,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct BatchFaucetResponse {
+    // This string is the Uuid for the req
+    pub task: Option<String>,
+    pub error: Option<String>,
+}
+
+#[derive(serde::Deserialize)]
+pub struct FaucetResponse {
+    task: Option<String>,
+    error: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct BatchStatusFaucetResponse {
+    pub status: Option<BatchSendStatus>,
+    pub error: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(rename_all = "UPPERCASE")]
+pub enum BatchSendStatusType {
+    Inprogress,
+    Succeeded,
+    Discarded,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BatchSendStatus {
+    pub status: BatchSendStatusType,
+    pub transferred_gas_objects: Option<FaucetReceipt>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct FaucetReceipt {
+    pub sent: Vec<CoinInfo>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BatchFaucetReceipt {
+    pub task: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct CoinInfo {
+    pub amount: u64,
+    pub id: ObjectId,
+    pub transfer_tx_digest: TransactionDigest,
+}
+
+impl FaucetClient {
+    /// Construct a new `FaucetClient` with the given `reqwest::Client`. Defaults to Testnet
+    /// faucet.
+    pub(crate) fn new(inner: reqwest::Client, rpc: &Url) -> Self {
+        let faucet_url = if rpc.as_str() == LOCAL_HOST {
+            Url::parse(FAUCET_LOCAL_HOST).expect("Invalid faucet URL")
+        } else if rpc.as_str() == DEVNET_HOST {
+            Url::parse(FAUCET_DEVNET_HOST).expect("Invalid faucet URL")
+        } else {
+            Url::parse(FAUCET_TESTNET_HOST).expect("Invalid faucet URL")
+        };
```

Comment:
Why are you trying to match the url against a network?

Reply:
So that you do not need to call `.devnet()` or set up specifically a faucet URL if the public nodes are used.

Reply:
basically you can do 
```rust
  let client = Client::new_devnet();
  let coin = client.faucet().request_and_wait("address").await?;
  ```

Reply:
Then this should be done from the GraphQL client not the faucet client. its very confusing to pass something in and have it reject what you gave it and instead do something completely different.

Reply:
I'd also argue that you shouldn't be able to go from graphql client -> faucet client. It conflates the two together and they could be very confusing if you're not using a whitelisted network address. It would be much clearer to require someone build a graphql client and a faucet client separately

Reply:
That makes sense. This should probably be a separate crate, as you do not need a client to do a faucet request.